src/Proxy/DriveItemProxy.php

```diff
@@ -255,45 +257,86 @@ public function delete()
      * @param string $name
-     *        The name.
-     * @param string|resource|\GuzzleHttp\Psr7\Stream $content
-     *        The content.
+     *        The name of file(relative file path if in same directory (or) absolute file path).
      * @param array $options
      *        The options.
      *
      * @return DriveItemProxy
      *         The drive item created.
      *
-     * @todo Support name conflict behavior.
-     * @todo Support content type in options.
      */
-    public function upload($name, $content, array $options = [])
+    public function upload($name, array $options = [])
     {
         $name         = rawurlencode($name);
         $driveLocator = "/drives/{$this->parentReference->driveId}";
         $itemLocator  = "/items/{$this->id}";
-        $endpoint     = "$driveLocator$itemLocator:/$name:/content";
-
-        $body = $content instanceof Stream ?
-            $content
-            : Psr7\stream_for($content);
-
-        $response = $this
+        $uploadSessionEndpoint = "$driveLocator$itemLocator:/$name:/createUploadSession";
+        //include extended options with existing item for upload session
+        $item = array_merge(
+            array("name" => $name, 'description' => '' ,"@microsoft.graph.conflictBehavior" => "rename"), $options
+        );
+
+		//create an upload session for large files
+		$uploadSessionResponse = $this
             ->graph
-            ->createRequest('PUT', $endpoint)
-            ->addHeaders($options)
-            ->attachBody($body)
+            ->createRequest('POST', $uploadSessionEndpoint)
+            ->addHeaders(array('Content-Type' => 'application/json'))
+            ->attachBody(array("item" =>$item))
             ->execute();
-
-        $status = $response->getStatus();
-
-        if ($status != 201) {
-            throw new \Exception("Unexpected status code produced by 'PUT $endpoint': $status");
-        }
-
-        $driveItem = $response->getResponseAsObject(DriveItem::class);
-
-        return new self($this->graph, $driveItem);
+		
+		$uploadSessionResponseBody=$uploadSessionResponse->getBody();
+
+        if(is_array($uploadSessionResponseBody) && array_key_exists("uploadUrl", $uploadSessionResponseBody)) {
+
+            $uploadUrl = $uploadSessionResponseBody['uploadUrl'];
+
+			$handle = @fopen($name, "r");
+			if ($handle) {
+				$bytesRead=0;
+				$buffer = "";
+				$totalSize=filesize($name);
+				
+				while (!feof($handle)) {
+					$buffer = fread($handle, CHUNK_SIZE);
+					$contentLength = strlen($buffer);
+					ob_flush(); //flush output buffer high-level
+					flush(); //flush output buffer low-level
+					
+					//initialise bytes read, and interval range
+					$byteIntervalStart=$bytesRead;
+					$bytesRead += $contentLength;
+					$byteIntervalEnd=$bytesRead-1;
+					
+					//upload headers containing bytes Content length and range
+					$uploadSessionHeaders = array();
+					$uploadSessionHeaders['Content-Type'] = 'application/json';
```

Comment:
I think this `Content-Type` is wrong: this should be the actual MIME type of the uploaded file, which you may ask the user for via `$options`.

Reply:
Yeah I copy pasted header from url request. I forgot to remove that. As the doc mentioned [here](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_createuploadsession?view=odsp-graph-online), It recognizes only **Content-Length** and **Content-Range**. Rest it doesn't care about I guess as we are uploading only chunks of file stream. I'll get back to the content type for the chunks a bit later

And, I got used with array() declaration. Haven't used shorthand notation for a long. Happy to change it to that in the script.